packages/@dcl/inspector/src/components/tree/Tree.tsx

```diff
@@ -0,0 +1,126 @@
+import React, { useState } from 'react'
+import { useDrag, useDrop } from 'react-dnd'
+
+import './Tree.css'
+
+import { Tree, TreeType } from '../../utils/tree'
+import { Input } from './input'
+import { Controls } from './controls'
+import { Icon } from './icon'
+
+interface Props {
+  value: Tree
+}
+
+interface TreeProps {
+  tree: Tree
+  update: () => void
+}
+
+const getEditModeStyles = (active: boolean) => ({ display: active ? 'none' : 'block' })
+const getExpandStyles = (active: boolean) => ({ height: active ? 'auto' : '0', overflow: 'hidden', display: 'block' })
+
+const MemoTree = React.memo(TreeComponent)
+
+function TreeComponent({ tree, update }: TreeProps) {
+  const { value, childs, parent, expanded = false, type } = tree
+  const [editMode, setEditMode] = useState(false)
+  const [insertMode, setInsertMode] = useState<TreeType | undefined>(undefined)
+
+  const [, drag] = useDrag(
+    () => ({
+      type: 'tree',
+      item: tree
+    }),
+    [tree]
+  )
+
+  const [, drop] = useDrop(() => ({
+    accept: 'tree',
+    drop: (item: typeof tree, monitor) => {
+      if (monitor.didDrop()) return
+      tree.becomeParentOf(item)
+      update()
+    }
+  }))
+
+  const quitEditMode = () => setEditMode(false)
+  const quitInsertMode = () => setInsertMode(undefined)
+
+  const handleToggleExpand = (_: React.MouseEvent) => {
+    tree.toggleExpand()
+    update()
+  }
+
+  const handleToggleEdit = (e: React.MouseEvent) => {
+    e.stopPropagation()
+    setEditMode(true)
+  }
+
+  const onChangeEditValue = (newValue: string) => {
+    tree.rename(newValue)
+    setEditMode(false)
+    update()
```

Comment:
this is an issue that i see in all the handlers, we are mutating instead of informing of the change intended to the parent and letting the parent perform the update to the state. 

Having that mutation is what forces you to have this `update()` call to force a re-render, but if instead the component just comunicated the action, then the parent can return a new tree which should in turn render it again organically.

For example, instead of performing the `tree.rename()` above, we should only call a callback informing the id of the node and the new label, like:

```
onRename(node.id, newLabel)
```

Then the parent is who decides how to handle that change and generate a new tree, otherwise this component can only work with the `Tree` class, which is not where the state will live.